builtin/docker/pull/kaniko.go

```diff
@@ -41,98 +41,177 @@ func (b *Builder) pullWithKaniko(
+		}
 
-	refPath := reference.Path(ref)
+		refPath := reference.Path(ref)
 
-	err = oci.Negotiate(ref.Name())
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to negotiate with upstream")
+		err = destinationProxy.Negotiate(ref.Name())
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to negotiate with upstream")
+		}
+
+		if !b.config.DisableCEB {
+			step.Update("Injecting entrypoint...")
+			// For Kaniko we can use our runtime arch because the image we build
+			// always matches the architecture of our Kaniko environment.
+			assetName, ok := assets.CEBArch[runtime.GOARCH]
+			if !ok {
+				return nil, status.Errorf(codes.FailedPrecondition,
+					"automatic entrypoint injection not supported for architecture: %s", runtime.GOARCH)
+			}
+
+			data, err := assets.Asset(assetName)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "unable to restore custom entrypoint binary: %s", err)
+			}
+			step.Done()
+
+			step = sg.Add("Testing registry and uploading entrypoint layer")
+			err = destinationProxy.SetupEntrypointLayer(refPath, data)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "error setting up entrypoint layer to host: %q, err: %s", destinationProxy.Upstream, err)
+			}
+			step.Done()
+		}
+
+		// Setting up local registry to which Kaniko will push
+		li, err := net.Listen("tcp", "localhost:0")
+		if err != nil {
+			return nil, err
+		}
+
+		defer li.Close()
+		go http.Serve(li, &destinationProxy)
+
+		port := li.Addr().(*net.TCPAddr).Port
+
+		localRef = fmt.Sprintf("localhost:%d/%s:%s", port, refPath, ai.Tag)
 	}
 
-	if !b.config.DisableCEB {
-		step.Update("Injecting entrypoint...")
-		// For Kaniko we can use our runtime arch because the image we build
-		// always matches the architecture of our Kaniko environment.
-		assetName, ok := assets.CEBArch[runtime.GOARCH]
-		if !ok {
-			return nil, status.Errorf(codes.FailedPrecondition,
-				"automatic entrypoint injection not supported for architecture: %s", runtime.GOARCH)
+	// sourceProxy is the OCI Proxy server that the Kaniko process will pull the
+	// base docker container from. It authenticates and proxies the image from
+	// the location specified in the build->docker-pull block.
+	var sourceProxy ociregistry.Server
+	// remoteRef represents the address of the pullProxy that will be inserted
+	// into an on-demand Docker file: ex "FROM localhost:1234/image:tag"
+	var remoteRef string
+	// DisableEntrypoint shouldn't matter on the source, but we set it to be
+	// consistent with the destination proxy
+	sourceProxy.DisableEntrypoint = b.config.DisableCEB
+	sourceProxy.Logger = log
+	{
+		if b.config.EncodedAuth != "" {
+			//If EncodedAuth is set, use that
+			user, pass, err := wpdocker.CredentialsFromConfig(b.config.EncodedAuth)
+			if err != nil {
+				return nil, err
+			}
+			sourceProxy.AuthConfig.Username = user
+			sourceProxy.AuthConfig.Password = pass
+		} else if *b.config.Auth != (wpdocker.Auth{}) {
```

Comment:
If `Auth` is nil, this will panic I believe.

Reply:
Added a nil check in e5b5fe97c